### DISCUSSION TITLE
Return Alpine proxy data object

### DISCUSSION THREAD

**Discussion by asbjornu** [alpinejs/alpine#4375] [3. Feature ideas]
As with #203, I would like to interact with an Alpine data component through out-of-scope JavaScript. I know I can use events to do communicate between the outside and inside, but as this particular case is a JavaScript class holding some state, which given some external conditions, eventually initializes an Alpine component, that's a bit cumbersome. All the communication would be between the outer class and the inner Alpine Data component, after all.

I could of course merge all of the surrounding code into the Alpine component, but that makes the interaction with the JavaScript class from both itself and from the outside very awkward. Not being able to use private fields,  `static` methods, etc. is very limiting. 

While following the Alpine.js best practice usually works fine, being able to interact with an Alpine data component more ergonomically from JavaScript is highly desirable.

There are currently at least two hacks available to get around this. One is `document.querySelector('["x-data"]').__x` and another is to wrap the initialization of Alpine in a `Promise`:

```js
function initializeFoo() {
    return new Promise((resolve) => {
        document.addEventListener('alpine:init', () => {
            Alpine.data('Foo', () => ({
                init() {
                    resolve(this);
                }
            }));
        });
    });
}

const foo = await initializeFoo();
```

An imho better option would be for the `Alpine.data()` method to return the proxied data object. As it currently returns `undefined`, that would be a non-breaking change that would remove the need to use the hacks described above.

**Comment by ekwoka** [comment]
This wouldn't work, purely because `Alpine.data` does not initialize a component.

You pass a component builder. The component is then initialized every time that data function is used in your document.

If you want one global one like this, then `Alpine.store` already works for this.

https://alpinejs.dev/globals/alpine-store

Is there a reason why Alpine store doesn't work for you?

**Reply by asbjornu** [reply]
> This wouldn't work, purely because `Alpine.data` does not initialize a component.

Right, of course. The data component is not initialized until it's used in `x-data`. Makes sense.

> Is there a reason why Alpine store doesn't work for you?

I try to avoid stores because they are globally available, encouraging outside manipulation and cross component reuse instead of local scope and events for messaging. I really like the fact that Alpine data components are scoped to a particular HTML element and its descendants, but in some cases like this, it would be very useful if it was a bit easier to interact with through JavaScript.

Perhaps something like `Alpine.getProxy($el)` could work instead, where it would in practice just return `$el.__x`, but through a public interface hiding `__x` as an implementation detail?

**Reply by ekwoka** [reply]
There is already `Alpine.$data(el)` which will do that as well.

https://awesomealpine.com/posts/underrated-alpine-api/#bonus-alpinedata

Or you can make a reactive proxy with `Alpine.reactive` and pass that however.

Or initialize your class from within the alpine component, since it seems you want it to be fairly local, or even use the Class directly as the data.

`Alpine.data('comp', () => new MyClass())` for example.

Hard to be more specific without more details. It seems you want something global, but also not global? It's a bit unclear.

**Reply by asbjornu** [reply]
> `Alpine.data('comp', () => new MyClass())` 

This might be a good option here. I'll try that out, thanks!

> It seems you want something global, but also not global? 

I don't want anything global, I just want to be able to manipulate the Alpine data component from a JavaScript class, both of which are logically scoped to the same HTML element. As there's no way to enforce the scope of arbitrary JavaScript to an HTML element like this, I write "logically scoped", as the business logic within the class only deals with the local scope of a particular HTML element.

**Reply by ekwoka** [reply]
Right. The simplest is just having Alpine instantiate the class, no?

Either as a property, or as the component entirely.

**Reply by asbjornu** [reply]
I tried `Alpine.data('comp', () => new MyClass())` now, and while it works, it requires the class to be highly aware of it being used as an Alpine proxy. I can't just reference `this.property.method()` everywhere anymore; I have to unwrap it with `Alpine.raw(this.property)` before invoking `method()` on the unwrapped object. If I don't unwrap, the following error occurs:

```js
Uncaught TypeError: Cannot read from private field
    __typeError InnerClass.js:2
    __accessCheck InnerClass.js:4
    __privateGet InnerClass.js:5
    pause InnerClass.js:228
    init MyClass.js:46
    setTimeout handler*init MyClass.js:44
    generateEvaluatorFromFunction alpinejs.js:463
    tryCatch alpinejs.js:416
    evaluate alpinejs.js:443
    <anonymous> alpinejs.js:2929
    flushHandlers alpinejs.js:587
    stopDeferring alpinejs.js:592
    deferHandlingDirectives alpinejs.js:595
    initTree alpinejs.js:785
    start alpinejs.js:737
    start alpinejs.js:736
    <anonymous> MyClass.js:27
    <anonymous> MyClass.js:29
InnerClass.js:2:9
    __typeError InnerClass.js:2
    __accessCheck InnerClass.js:4
    __privateGet InnerClass.js:5
    pause InnerClass.js:228
    init MyClass.js:46
    (Async: setTimeout handler)
    init MyClass.js:44
    InterpretGeneratorResume self-hosted:1420
    AsyncFunctionNext self-hosted:807
    (Async: async)
    generateEvaluatorFromFunction module.esm.js:477
    tryCatch module.esm.js:428
    evaluate module.esm.js:457
    <anonymous> module.esm.js:3033
    flushHandlers module.esm.js:603
    stopDeferring module.esm.js:608
    deferHandlingDirectives module.esm.js:611
    initTree module.esm.js:809
    start module.esm.js:761
    forEach self-hosted:160
    start module.esm.js:760
    <anonymous> main.js:27
    <anonymous> main.js:29

```

I also can't use [private properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties) on the class within methods invoked by Alpine, such as `init()`. If I do something like `this.#property = 'some value';` within the `init()` method, the following error occurs:

```js
Uncaught (in promise) TypeError: can't set private field: object is not the right class
    init MyClass.js:39
    generateEvaluatorFromFunction alpinejs.js:463
    tryCatch alpinejs.js:416
    evaluate alpinejs.js:443
    <anonymous> alpinejs.js:2929
    flushHandlers alpinejs.js:587
    stopDeferring alpinejs.js:592
    deferHandlingDirectives alpinejs.js:595
    initTree alpinejs.js:785
    start alpinejs.js:737
    start alpinejs.js:736
    <anonymous> main.js:27
    <anonymous> main.js:29
```

I thus have to declare all properties as public if I want to access them from an Alpine context. Thoughts?

**Reply by ekwoka** [reply]
Just make them public.

Its your code. Why bother making it runtime private?

**Reply by asbjornu** [reply]
Marking fields and methods as private acts as documentation of how the class is supposed to be used. It guides other developers in how to interact with the class as it was designed. But even making every field and method in the class public, I have difficulties getting the state propagated properly between the innards of the class and the Alpine data proxy.

You can see the issue I'm having in [the following CodePen](https://codepen.io/asbjornu/pen/PoMqNdQ?editors=1010). Every click on a new link in the list of links should increment the number, but as you can see, Alpine's state is not in sync with that of the class other than for the `@click` event. Any help resolving this would be highly appreciated! 🙏🏼

**Reply by SimoTod** [reply]
If you use this in the constructor, you are accessing the real instance of the class. Alpine reacrivity works because it creates proxies around properties so it can detect when they change.

It should work if you do

```javascript
class Incrementer {
  constructor() {
    this.number = 0;
    this.increment({ type: 'constructor' })
  }

  init() {
 window.addEventListener('hashchange', (event) => this.increment(event), false);
  }
...
```

**Reply by ekwoka** [reply]
> Marking fields and methods as private acts as documentation of how the class is supposed to be used. It guides other developers in how to interact with the class as it was designed.

If you're using TypeScript, you can do this with the `public` and `private` keywords.

A lot depends on what your actual goals are, so the most I and @SimoTod can do is present multiple approaches.

Highly reduced examples of what issues you're having don't really help since most of them will fail to provide a compelling reason why they should be done that way at all.

I've made a lot of very capable, adaptive components and not really run into these issues, so it might be a major XY problem. Where you settled on Y as the way to do X, and are having issues, but Y just wasn't a good solution in the first place.

**Reply by asbjornu** [reply]
@SimoTod, ah, that makese sense – thanks! Everything now seems to work. Although I've been forced to make everything in the class public, I suppose that's an ok compromise to make.

> If you're using TypeScript, you can do this with the public and private keywords.

True, but I'm not using TypeScript, just regular JavaScript.

> I've made a lot of very capable, adaptive components and not really run into these issues, so it might be a major XY problem. Where you settled on Y as the way to do X, and are having issues, but Y just wasn't a good solution in the first place.

That might indeed be the case, but for now it seems like using a JavaScript `class` with public members is a good compromise. I'll probably refactor towards other patterns over time as I understand Alpine.js better. Thanks for the help – much appreciated! 🙏🏼
